wgpu-native/src/track/range.rs

```diff
@@ -0,0 +1,450 @@
+use std::{
+    cmp::Ordering,
+    fmt::Debug,
+    iter::Peekable,
+    ops::Range,
+    slice::Iter,
+};
+
+/// Structure that keeps track of a I -> T mapping,
+/// optimized for a case where keys of the same values
+/// are often grouped together linearly.
+#[derive(Clone, Debug)]
+pub struct RangedStates<I, T> {
+    ranges: Vec<(Range<I>, T)>,
+}
+
+impl<I, T> Default for RangedStates<I, T> {
+    fn default() -> Self {
+        RangedStates {
+            ranges: Vec::new(),
+        }
+    }
+}
+
+impl<I: Copy + PartialOrd, T: Copy + PartialEq> RangedStates<I, T> {
+    /// Construct a new instance from a slice of ranges.
+    #[cfg(test)]
+    pub fn new(values: &[(Range<I>, T)]) -> Self {
+        RangedStates {
+            ranges: values.to_vec(),
+        }
+    }
+
+    /// Clear all the ranges.
+    pub fn clear(&mut self) {
+        self.ranges.clear();
+    }
+
+    /// Append a range.
+    pub fn append(&mut self, index: Range<I>, value: T) {
+        if let Some(last) = self.ranges.last() {
+            debug_assert!(last.0.end <= index.start);
+        }
+        self.ranges.push((index, value));
+    }
+
+    /// Iterate through the stored ranges immutably.
+    pub fn iter(&self) -> Iter<(Range<I>, T)> {
+        self.ranges.iter()
+    }
+
+    /// Check that all the ranges are non-intersecting and ordered.
+    /// Panics otherwise.
+    #[cfg(test)]
+    fn check_sanity(&self) {
+        for a in self.ranges.iter() {
+            assert!(a.0.start < a.0.end);
+        }
+        for (a, b) in self.ranges.iter().zip(self.ranges[1..].iter()) {
+            assert!(a.0.end <= b.0.start);
+        }
+    }
+
+    /// Merge the neighboring ranges together, where possible.
+    #[cfg(test)]
+    fn coalesce(&mut self) {
```

Comment:
The idea is that these methods are used by regular operation of the tracker, it's just that I haven't found a good place to call them from, yet. the `cfg(test)` is here pretty much to avoid a compile warning, nothing more